dhis-2/dhis-services/dhis-service-node/src/test/java/org/hisp/dhis/node/serializer/CsvNodeSerializerTest.java

```diff
@@ -0,0 +1,193 @@
+        assertEquals( "\"First simple node\"", values.get( 0 ).get( 0 ) );
+        assertEquals( "\"Second value simple child\"", values.get( 1 ).get( 1 ) );
+        assertEquals( "\"Third simple node\"", values.get( 0 ).get( 2 ) );
+    }
+
+    @Test
+    void CsvFileIsWrittenWhenAttributesAreProvided()
+        throws Exception
+    {
+        csvNodeSerializer.serialize( createComplexCollection( "attributes", true ),
+            new FileOutputStream( "output.csv" ) );
+
+        List<List<String>> values = readTestFile();
+
+        assertEquals( 2, values.size() );
+        assertEquals( 6, values.get( 0 ).size() );
+        assertEquals( 6, values.get( 1 ).size() );
+        assertEquals( "\"First simple node\"", values.get( 0 ).get( 0 ) );
+        assertEquals( "\"Second attr\"", values.get( 0 ).get( 3 ) );
+        assertEquals( "\"Fourth attr\"", values.get( 0 ).get( 5 ) );
+        assertEquals( "\"First attr value\"", values.get( 1 ).get( 2 ) );
+        assertEquals( "\"Third attr value\"", values.get( 1 ).get( 4 ) );
+    }
+
+    @Test
+    void CsvFileIsWrittenWhenSimpleNodesAreProvidedButAttributeComplexNodeIsNot()
+        throws Exception
+    {
+        csvNodeSerializer.serialize( createComplexCollection( "enrollments", true ),
+            new FileOutputStream( "output.csv" ) );
+
+        List<List<String>> values = readTestFile();
+
+        assertEquals( 2, values.size() );
+        assertEquals( 2, values.get( 0 ).size() );
+        assertEquals( 2, values.get( 1 ).size() );
+        assertEquals( "\"First simple node\"", values.get( 0 ).get( 0 ) );
+        assertEquals( "\"Second value simple child\"", values.get( 1 ).get( 1 ) );
+    }
+
+    @Test
+    void CsvFileIsNotWrittenWhenNoSimpleNodesNorAttributeComplexNodeAreProvided()
+        throws Exception
+    {
+        Exception exception = assertThrows( CsvWriteException.class, () -> csvNodeSerializer
+            .serialize( createComplexCollection( "enrollments", false ), new FileOutputStream( "output.csv" ) ) );
+
+        String expectedMessage = "Schema specified that header line is to be written; but contains no column names";
+        String actualMessage = exception.getMessage();
+
+        assertTrue( actualMessage.contains( expectedMessage ) );
+    }
+
+    @AfterEach
+    public void cleanUp()
+    {
+        File f = new File( "output.csv" );
+        if ( !f.delete() )
+        {
+            log.warn( "Not possible to delete file {}", f.getName() );
+        }
+    }
+
+    private RootNode createCollectionWithSimpleNodes()
+    {
+        CollectionNode collectionRootNode = new CollectionNode( "Complex node" );
+        CollectionNode firstCollectionNode = new CollectionNode( "First collection child" );
+        CollectionNode secondCollectionNode = new CollectionNode( "Second collection child" );
+
+        SimpleNode firstSimpleNode = new SimpleNode( "First simple node", "First value simple child" );
```

Comment:
I would inline vars like `firstSimpleNode` since they are only used once. This makes me have to read/remember less IMHO 😄